aQute.libg/src/aQute/lib/memoize/CloseableMemoizingSupplier.java

```diff
@@ -2,74 +2,109 @@
 
 import static java.util.Objects.requireNonNull;
 
+import java.util.concurrent.locks.StampedLock;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 
+/**
+ * The object can exist in one of three states:
+ * <ul>
+ * <li>initial which means neither {@code get} or {@code close} has been called.
+ * From this state, the object can transition directly to either open or
+ * closed.</li>
+ * <li>open which means memoized is non-null with the value from the wrapped
+ * supplier. From this state, the object can transition to closed.</li>
+ * <li>closed which means memoized is null and any value it may have held was
+ * closed. This is a terminal state.</li>
+ * </ul>
+ */
 class CloseableMemoizingSupplier<T extends AutoCloseable> implements CloseableMemoize<T> {
 	private final Supplier<? extends T>	supplier;
-	// @GuardedBy("this")
-	private boolean						closed	= false;
-	private volatile T					memoized;
+	private final StampedLock			lock;
+	private volatile boolean			initial;
+	// @GuardedBy("initial")
+	private T							memoized;
 
 	CloseableMemoizingSupplier(Supplier<? extends T> supplier) {
 		this.supplier = requireNonNull(supplier);
+		lock = new StampedLock();
+		initial = true;
 	}
 
 	@Override
 	public T get() {
-		if (memoized == null) {
-			synchronized (this) {
-				if (closed)
-					throw new IllegalStateException("Already closed");
-				return get0();
+		// read initial _before_ read memoized
+		if (initial) {
+			// critical section: only one at a time
+			final long stamp = lock.writeLock();
+			try {
+				if (initial) {
+					T result = requireNonNull(supplier.get());
+					memoized = result;
+					// write initial _after_ write memoized
+					initial = false;
```

Comment:
The supplier violates the contract by returning null. Unless you want to have the supplier returning null result in closing the memoizing supplier.

Reply:
Well, why not just call it once to at least obey the contract?

Reply:
If it returns null, we could declare that to be a close of the memoizeing supplier. So the memoizing supplier would enter the closed state when the wrapper supplier returns null. If that in line with what you are thinking?

Reply:
We still then have the problem of what to do with the caller to our get method. Is the null then returned to him? The current state means he will get an NPE to at least alert him to the fact that the wrapped supplier returned a null. Otherwise, the caller may need to null check the result from our get.

Reply:
If we treat the wrapper supplier returning null as causing the memoized supplier to be closed, then the proper thing would be to throw an IllegalStateException since that is what would have happened if get was called on a closed memoized supplier.

I made #3973 to show this.